### PR TITLE
Ensure lock state is always valid

### LINKED ISSUE #55: MaxConcurrency does not work
This is going to be vague, as I'm currently trying to debug what's going on, but the `MaxConcurrency` option seems to be broken.

*According to my current understanding*:

For example, the `:lock` key for a job, under load, should be the same as the value of `MaxConcurrency`, but mine stays at 0.
The `:lock_info` key should have individual lock counts, per worker, that add up to the total of `:lock` for that job. However, the contents of that hash vary wildly between `-4` and `4`. I am testing with a `MaxConcurrency` of 1.

It seems there is a race condition in the increment/decrement logic. I have a worker pool size of 4, so the -4 and 4 variance could come from there. Perhaps the system is attempting to pull 4 jobs at once, and all 4 succeed in incrementing/decrementing?

### PR DESCRIPTION
**Pull request by tylerb**
Fix #55 

----

This PR introduces functionality to ensure that the lock state is always valid for a given `maxConcurrency` value. It does so by checking to see if `lock <0 or lock > maxConcurrency`. If that state is detected, the entire lock system for that job is reset so that we can begin processing jobs from a clean lock state.

In addition, we no longer increment jobs when `maxConcurrency == 0`. There is no need to acquire a lock when we are not restricting the job to a specific max concurrency.

### CODE REVIEW

File: redis.go [hunk 1]
```diff
@@ -118,10 +143,14 @@ local function isPaused(pauseKey)
   return redis.call('get', pauseKey)
 end
 
-local function canRun(lockKey, maxConcurrency)
+local function canRun(lockKey, lockInfoKey, maxConcurrency)
+  if (not maxConcurrency or maxConcurrency == 0) then
```
Comment (tylerb) [inline redis.go]: This was moved to the top of the function to remove an unnecessary call to redis.

**Comment by tylerb** [comment]
I'd love to get this merged soon. Thanks!

**Comment by hoffoo** [comment]
@shdunning can you check this out?

**Comment by shdunning** [comment]
closing this out as pr #58 addresses the original concern.

**Pull request closed without merging**